components/rich_text_editor/extensions/link/utils.js

```diff
@@ -0,0 +1,189 @@
+import { getMarksBetween } from '@tiptap/core';
+import {
+  getPhoneNumberRegex,
+  linkRegex,
+} from '@/common/utils';
+
+/**
+ * Get matches in a string and return the ones that pass the extra validation.
+ */
+export function getRegexMatches (text, regex, validator) {
+  const matches = [];
+
+  // Reset the lastIndex since the last time this was run.
+  regex.lastIndex = 0;
+
+  let match;
+  while ((match = regex.exec(text))) {
+    if (validator(text, match)) {
+      matches.push(match);
+    }
+  }
+
+  return matches;
+}
+
+/**
+ * Validate the prefix of a match in a string not to contain certain characters.
+ */
+export function hasValidPrefix (text, match) {
+  // The string match can't start with # or @ or have either preceding the match
+  // as they're reserved for mentions and hashtags.
+  return !['#', '@'].includes(text.charAt(match.index)) &&
+    !['#', '@'].includes(text.charAt(match.index - 1));
+}
+
+/**
+ * Trim punctuation characters at the a of the string, e.g. "dialpad.com!" =>
+ * "dialpad.com"
+ */
+export function trimEndPunctiation (string) {
+  const endPunctuationRegex = new RegExp(
+    '(?:' +
+    [
+      '[!?.,:;\'"]',
+      '(?:&|&amp;)(?:lt|gt|quot|apos|raquo|laquo|rsaquo|lsaquo);)+$',
+    ].join('|'),
+    'g',
+  );
+  return string.replace(endPunctuationRegex, '');
+}
+
+/**
+ * Find the word from a string at a given index. For example for "example here"
+ * - indices 0-7  => "example"
+ * - indices 8-12 => "here".
+ * Modified from https://stackoverflow.com/a/5174867
+ */
+export function getWordAt (text, index) {
+  // Position of the first non-whitespace character following a possible
+  // whitespace when looking from the index to the left.
+  const left = text.slice(0, index + 1).search(/\S+\s*$/);
+
+  // Position of the first whitespace when looking from the index to the right.
+  const right = text.slice(index).search(/\s/);
+
+  // If the word is at the end of the string, right is -1.
+  if (right < 0) {
+    const word = text.slice(left);
+    return {
+      text: word,
+      from: left,
+      to: left + word.length,
+    };
+  }
+
+  return {
+    text: text.slice(left, right + index),
+    from: left,
+    to: right + index,
+  };
+}
+
+/**
+ * Helper to check if a word at a given index matches a regex and if true, finds
+ * the previous or next word until the regex doesn't match anymore. Useful to
+ * find for example the entire phone number when it is separated by whitespace.
+ */
+export function getWordAtUntil (text, index, direction, regex) {
+  const word = getWordAt(text, index);
+
+  // If the word doesn't match the regex we can just return it.
+  if (!regex.test(word.text)) {
+    return word;
+  }
+
+  // Depending on the direction take one step to the left or right to find the
+  // preceding or following word.
+  const newIndex = direction === 'left' ? word.from - 1 : word.to + 1;
+
+  // Prevent an infinite loop for the edge cases.
```

Comment:
sorry to be a terminology snob but I believe the correct term is "base cases" in a recursive function.